File: demo/web_site/src/types/api/leaderboard.ts
/**
 * 排行榜相关API类型定义
 */

// ==================== 排行榜类型 ====================

/**
 * 排名类型
 */
export type RankType = 'strategy' | 'signal'

/**
 * 统计周期
 */
export type Period = 'daily' | 'weekly' | 'monthly' | 'all_time'

/**
 * 排序字段
 */
export type SortBy = 'total_return' | 'sharpe_ratio' | 'win_rate' | 'annual_return'

/**
 * 排行榜条目
 */
export interface LeaderboardApiEntry {
  id: string
  rank_type: RankType
  period: Period
  rank_position: number
  entity_id: string
  entity_name: string
  entity_type: string
  owner_id: string | null
  owner_name: string | null
  total_return: number | null
  annual_return: number | null
  max_drawdown: number | null
  sharpe_ratio: number | null
  win_rate: number | null
  total_trades: number | null
  profit_factor: number | null
  stat_start_time: string | null
  stat_end_time: string | null
  snapshot_time: string | null
}

/**
 * 综合排行榜响应
 */
export interface LeaderboardOverviewResponse {
  period: Period
  snapshot_time: string | null
  strategy_ranking: LeaderboardApiEntry[]
  signal_ranking: LeaderboardApiEntry[]
}

/**
 * 排行榜列表响应
 */
export interface LeaderboardResponse {
  items: LeaderboardApiEntry[]
  total: number
  period: Period
  sort_by?: SortBy
  snapshot_time: string | null
}

/**
 * 综合排行榜查询参数
 */
export interface LeaderboardOverviewParams {
  period?: Period
  limit?: number
}

/**
 * 策略排行榜查询参数
 */
export interface StrategyLeaderboardParams {
  period?: Period
  sort_by?: SortBy
  limit?: number
}

/**
 * 信号排行榜查询参数
 */
export interface SignalLeaderboardParams {
  period?: Period
  limit?: number
}

/**
 * 刷新排行榜响应
 */
export interface RefreshLeaderboardResponse {
  success: boolean
  message: string
  snapshot_time: string
}

File: demo/quant_trading_system/src/quant_trading_system/models/leaderboard.py
"""
排行榜模块数据库模型
"""

from datetime import datetime

from sqlalchemy import Column, String, DateTime, Integer, BigInteger, Numeric

from quant_trading_system.core.snowflake import generate_snowflake_id
from quant_trading_system.models.base import Base


class LeaderboardSnapshot(Base):
    """排行榜快照模型"""
    __tablename__ = "leaderboard_snapshots"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    rank_type = Column(String(32), nullable=False)
    period = Column(String(16), nullable=False)
    rank_position = Column(Integer, nullable=False)
    entity_id = Column(String(128), nullable=False)
    entity_name = Column(String(256))
    entity_type = Column(String(64))
    owner_id = Column(BigInteger)
    owner_name = Column(String(128))
    total_return = Column(Numeric(10, 6))
    annual_return = Column(Numeric(10, 6))
    max_drawdown = Column(Numeric(10, 6))
    sharpe_ratio = Column(Numeric(10, 6))
    win_rate = Column(Numeric(10, 6))
    total_trades = Column(Integer, default=0)
    profit_factor = Column(Numeric(10, 4))
    stat_start_time = Column(DateTime)
    stat_end_time = Column(DateTime)
    snapshot_time = Column(DateTime, default=datetime.utcnow)
    created_at = Column(DateTime, default=datetime.utcnow)
